include/sys/onoff.h

```diff
@@ -560,6 +561,73 @@ int onoff_monitor_register(struct onoff_manager *mgr,
 int onoff_monitor_unregister(struct onoff_manager *mgr,
 			     struct onoff_monitor *mon);
 
+/**
+ * @brief State used when a driver uses the on-off service API for synchronous
+ * operations.
+ *
+ * This is useful when a subsystem API uses the on-off API to support
+ * asynchronous operations but the transitions required by a particular driver
+ * are isr-ok and not sleep.
+ */
+struct onoff_sync_service {
+	/* Mutex protection for other fields. */
+	struct k_spinlock lock;
+
+	/* Negative is error, non-negative is reference count. */
+	int32_t count;
+};
+
+/**
+ * @brief Lock a synchronous onoff service and provide its state.
+ *
+ * @note If an error state is returned it is the caller's responsibility to
+ * decide whether to preserve it (finalize with the same error state) or clear
+ * the error (finalize with a non-error result).
+ *
+ * @param srv pointer to the synchronous service state.
+ *
+ * @param keyp pointer to where the lock key should be stored
+ *
+ * @return negative if the service is in an error state, otherwise the number of
+ * active requests at the time the lock was taken.
+ */
+int onoff_sync_lock(struct onoff_sync_service *srv,
+		    k_spinlock_key_t *keyp);
+
+/**
+ * @brief Process the completion of a transition in a synchronous service.
```

Comment:
`... and unlocks the context` or something similar.

Reply:
Part of processing is releasing the lock; it's mentioned in the first sentence after the brief.  I'll rework to address the other comments.

Reply:
my bad, in fact it is mentioned. However, `onoff_sync_lock` mentions that in the brief section.